Allow Chromecast provider to handle/avoid images which are "too big"

ijc opened the pull request:
My Chromecast Ultra crashes when presented with cover images on the order of 5000x5000 pixels (I've not determined the actual limit, 2048x2048 is ok though).

Allow users to avoid this problem by providing an advanced option per player which downscales the image to a specified size.

Fixes: https://github.com/music-assistant/hass-music-assistant/issues/3285

ijc committed:
Feat: setup.sh: Allow overriding which Python binary to use for venv

Debian systems do not have Python3 as `python`. With this change developers on
that platform can use:

```console
$ PYTHON=python3 ./scripts/setup.sh
```

marcelveldt commented:
I don't think this should even be an option. Let's just limit all images that are sent to a player directly to 512x512, which is also the max for some other players like airplay

ijc commented:
My logic was that this was in the same set of things as the "force mp3" behaviour, but that's fine, I can make the change. Do you want this to happen in the core `PlayerQueuesController` level or per-player provider?

marcelveldt commented:
> My logic was that this was in the same set of things as the "force mp3" behaviour, but that's fine, I can make the change. Do you want this to happen in the core `PlayerQueuesController` level or per-player provider?

Yeah, but enabling the mp3 mode is more a choice where this is more a bug; sending a large image will simply crash the device. 

Let's just adjust the "get_image_url" with a "size=512" parameter in the individual player providers, starting with cast.


![Scherm­afbeelding 2024-12-19 om 17 29 34](https://github.com/user-attachments/assets/a7f652e2-30e3-4fe3-ad6a-2d53f92af389)

ijc commented:
Ah, I misunderstood the mp3 option as being a workaround for buggy players too.

> ... adjust the "get_image_url" with a "size=512" parameter in the individual player providers,...

Changing just that one in `update_flow_metadata` is not sufficient IME.

The one in `PlayerQueuesController.player_media_from_queue_item` ([here](https://github.com/music-assistant/server/blob/73545eef0d98383d79c89ee9a09bda586eeeba78/music_assistant/controllers/player_queues.py#L1264)) also needs to change, the URL from that one ends up in the per player `play_media` and `enqueue_next_media` hooks (as part of the `PlayerMedia` parameter) which is why I was patching in `_handle_limit_image_size` into those locations.

In this PR I didn't touch the one in `update_flow_metadata` at all and things work for me when playing back from the local provider. Looks like `update_flow_metadata` is only used for `queue.flow_mode` (which isn't used in chromecast player?) or for radio streams (which I've never tried).

ijc commented:
I added some radio stations and those do go via `update_flow_metadata` as expected -- none of the ones I tried actually had an image url, but if they did I'd expect them to be on the small side and not trigger the issue anyway.

marcelveldt commented on music_assistant/providers/chromecast/__init__.py:
```diff
@@ -624,19 +624,47 @@ async def _disconnect_chromecast(self, castplayer: CastPlayer) -> None:
         castplayer.status_listener = None
         self.castplayers.pop(castplayer.player_id, None)
 
+    def _clamp_image_size(self, limit: int, orig_url: str | None) -> str | None:
+        """Clamp image for media item."""
+        if not orig_url or limit == 0:
+            return orig_url
+
+        try:
+            from urllib.parse import parse_qsl, urlencode, urlparse
+
+            url = urlparse(orig_url)
+            qsl = dict(parse_qsl(url.query))
+
+            size = int(qsl.get("size", "0"))
+            if size != 0 and size <= limit:
+                return orig_url
+
+            qsl["size"] = str(limit)
+            url = url._replace(query=urlencode(qsl))
+            return url.geturl()
+        except Exception as err:
+            self.logger.warning(
+                "Error while clamping image url size: %s",
+                str(err),
+                exc_info=err if self.logger.isEnabledFor(10) else None,
+            )
+
+        return orig_url
```
This assumes that every image has a query string with a size parameter while that is strictly speaking not true. If you want to absolutely guarantee the image size, you will have to request the image through the image proxy, which can also do resizing. 

I'm ok with this change as a start but you will still run into the issue if you receive an URL to an image that has no size in a query string but is just a super large image

ijc replied:
If you would prefer I could make the change to the call to `get_image_url` in the `PlayerQueuesController` instead, that would catch all (?) URL generation but would be affecting all players not just Chromecast.

marcelveldt replied:
ha, yeah that makes a lot more sense - that is a single line change and fixes it everywhere. Good call

ijc replied:
Done!

I changed the calls to `get_image_url` in `PlayerQueuesController` and the Chromecast player.

I didn't touch the calls in the airplay player (already sets size=500) or slimproto + sonos players (no known issues there). They will get the 512 limit from `PlayerQueuesController` but not in their own calls, that seems a bit inconsistent -- would it be better to add the 512 to those two as well while I'm here?

marcelveldt replied:
yeah, and always feel free to submit/propose optimizations when you see fit.
Thanks!

marcelveldt commented:
> Changing just that one in update_flow_metadata is not sufficient IME.

That was just a code example. It would need to be adjusted in all parts that handle the image from the queue item.
If you want to be 100% certain that a resized image is sent, we route everything through the image proxy.

Idea can be that we go with your patch now and if this pops up again, we know we have to adjust it to the proxy everywhere ?

ijc committed:
Fix: Limit images to 512x512

My Chromecast Ultra crashes when presented with cover images on the order of
5000x5000 pixels (I've not determined the actual limit, 2048x2048 is ok
though).

This sets the default size in the player queue to 512x512 which affects all
players, as well as a chromecast specific codepath. Of the other callers of
`get_image_url` airplay already sets a size of 500x500 while the slimproto and
sonos players do not set any limit.

Fixes: https://github.com/music-assistant/hass-music-assistant/issues/3285

marcelveldt reviewed (approved):
Thanks!
